### PR TITLE
Add Cursor agent support on new agent interface

### PR DESCRIPTION
**Pull request by squishykid**
## Summary

Adds Cursor as a supported agent with full hook lifecycle, transcript storage, and checkpoint integration.

### New: Cursor agent package (`agent/cursor/`)
- `types.go` — Cursor-specific types: session info, prompt input, pre-compact, subagent start/stop with native Cursor fields (`conversation_id`, `subagent_id`, `task`, etc.)
- `cursor.go` — Agent implementation with JSONL transcript storage (ReadSession, WriteSession, ChunkTranscript, ReassembleTranscript)
- `lifecycle.go` — `ParseHookEvent` mapping 7 Cursor hooks (`session-start`, `session-end`, `before-submit-prompt`, `stop`, `pre-compact`, `subagent-start`, `subagent-stop`) to normalized lifecycle events, plus `ReadTranscript`
- `hooks.go` — Hook install/uninstall for `.cursor/hooks.json` with unknown field preservation
- `hooks_test.go` — 11 hook tests (install, idempotent, force, preserve existing/unknown fields, uninstall)
- `lifecycle_test.go` — 14 lifecycle tests covering all hook types, edge cases, conversation_id fallback, and ReadTranscript
- `cursor_test.go` — 32 tests covering identity, session storage (ReadSession/WriteSession round-trips), chunking, DetectPresence, and sanitizePathForCursor

### Modified
- `registry.go` — Adds `AgentNameCursor` and `AgentTypeCursor` constants
- `explain.go` — Includes `AgentTypeCursor` in JSONL transcript scoping/offset logic
- `hooks_cmd.go` — Blank import for cursor package `init()` registration
- `summarize/summarize.go` + `summarize_test.go` — Cursor agent support in summarization
- `transcript/types.go` — Documented `Role` field (Cursor uses `role`, Claude Code uses `type`)
- `transcript/parse.go` + `parse_test.go` — Added shared `ExtractModifiedFiles` utility for agents with tool_use blocks

### Design decisions
- **Native Cursor hook format**: Uses Cursor's own subagent fields (`subagent_id`, `task`, `subagent_type`, etc.) rather than mapping to Claude Code conventions
- **7 hooks installed**: `session-start`, `session-end`, `before-submit-prompt`, `stop`, `pre-compact`, `subagent-start`, `subagent-stop`
- **No TranscriptAnalyzer**: Cursor's transcript format does not contain `tool_use` blocks, so modified file detection relies on git status instead of transcript parsing
- **No TokenCalculator**: Cursor's transcript does not include token usage data
- **FormatResumeCommand**: Returns a human-readable instruction ("Open this project in Cursor...") since Cursor is a GUI IDE with no CLI resume command

## Test plan

- [x] Unit tests pass (`mise run test`)
- [x] Integration tests pass (`mise run test:integration`)
- [x] Lint clean (`mise run lint`)
- [x] Builds successfully (`go build ./cmd/entire/`)
- [ ] Manual: `entire enable --agent cursor` creates `.cursor/hooks.json`
- [ ] Manual: `entire disable` removes hooks from `.cursor/hooks.json`
- [ ] Manual: `entire hooks cursor --help` shows all hook verbs

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: cmd/entire/cli/agent/cursor/hooks.go [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+
+	"github.com/entireio/cli/cmd/entire/cli/agent"
+	"github.com/entireio/cli/cmd/entire/cli/jsonutil"
+	"github.com/entireio/cli/cmd/entire/cli/paths"
+)
+
+// Ensure CursorAgent implements HookSupport and HookHandler
+var (
+	_ agent.HookSupport = (*CursorAgent)(nil)
+	_ agent.HookHandler = (*CursorAgent)(nil)
```
Comment (squishykid) [inline cmd/entire/cli/agent/cursor/hooks.go]: HookHandler seems deprecated, should use HookNames instead

File: cmd/entire/cli/agent/cursor/lifecycle.go [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+	"os"
+	"time"
+
+	"github.com/entireio/cli/cmd/entire/cli/agent"
+	"github.com/entireio/cli/cmd/entire/cli/textutil"
+	"github.com/entireio/cli/cmd/entire/cli/transcript"
+)
+
+// Compile-time interface assertions for new interfaces.
+var _ agent.TranscriptAnalyzer = (*CursorAgent)(nil)
```
Comment (squishykid) [inline cmd/entire/cli/agent/cursor/lifecycle.go]: need to check cursor implenents TranscriptAnalyzer properly

File: cmd/entire/cli/agent/cursor/hooks.go (hunk 1 above)
Comment (squishykid) [reply cmd/entire/cli/agent/cursor/hooks.go]: Ok, seems like this is still required during the migration.

File: cmd/entire/cli/agent/cursor/cursor.go [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+	repoRoot, err := paths.RepoRoot()
+	if err != nil {
+		repoRoot = "."
+	}
+
+	cursorDir := filepath.Join(repoRoot, ".cursor")
+	if _, err := os.Stat(cursorDir); err == nil {
+		return true, nil
+	}
+	return false, nil
```
[the remark this answers is not in the document]
Comment (squishykid) [reply cmd/entire/cli/agent/cursor/cursor.go]: .cursor directory should be sufficient in this case

File: cmd/entire/cli/agent/cursor/lifecycle.go [hunk 4]
```diff
@@ -101,42 +98,40 @@ func (c *CursorAgent) parseSessionEnd(stdin io.Reader) (*agent.Event, error) {
...
+		SessionID:  raw.ConversationID,
 		SessionRef: raw.TranscriptPath,
 		ToolUseID:  raw.ToolUseID,
 		ToolInput:  raw.ToolInput,
 		Timestamp:  time.Now(),
 	}
-	if raw.ToolResponse.AgentID != "" {
-		event.SubagentID = raw.ToolResponse.AgentID
-	}
+	//  TODO "tool_output": "{\"status\":\"success\",\"agentId\":\"3211cc34-8d8f-42de-9dcf-c19625b17566\",\"durationMs\":7901,\"messageCount\":1,\"toolCallCount\":1}",
```
Comment (squishykid) [inline cmd/entire/cli/agent/cursor/lifecycle.go]: need to fix sub-agent extraction

File: cmd/entire/cli/agent/cursor/hooks.go [hunk 5]
```diff
@@ -0,0 +1,371 @@
...
+	_ agent.HookSupport = (*CursorAgent)(nil)
+)
+
+// Cursor hook names - these become subcommands under `entire hooks cursor`
+const (
+	HookNameSessionStart       = "session-start"
+	HookNameSessionEnd         = "session-end"
+	HookNameBeforeSubmitPrompt = "before-submit-prompt"
+	HookNameStop               = "stop"
+	HookNamePreTool            = "pre-tool"
```
Comment (squishykid) [inline cmd/entire/cli/agent/cursor/hooks.go]: use subagent hooks, not pre-tool

**Comment by squishykid** [comment]
@cursor review

**Comment by squishykid** [comment]
@cursor review

**Comment by squishykid** [comment]
@cursor review

File: cmd/entire/cli/strategy/manual_commit_condensation.go [hunk 6]
```diff
@@ -591,6 +591,11 @@ func extractUserPrompts(agentType agent.AgentType, content string) []string {
 // where the current checkpoint began, allowing calculation for only the portion
 // of the transcript since the last checkpoint.
 func calculateTokenUsage(agentType agent.AgentType, data []byte, startOffset int) *agent.TokenUsage {
+	// No token usage information from Cursor yet
```
[the remark this answers is not in the document]
Comment (squishykid) [reply cmd/entire/cli/strategy/manual_commit_condensation.go:594]: Cursor does not implement TranscriptAnalyzer yet, this will come later @copilot

File: cmd/entire/cli/strategy/manual_commit_condensation.go
Comment (Soph) [inline cmd/entire/cli/strategy/manual_commit_condensation.go:545]: We have special handling for OpenCode and Gemini her, is the format just the same for Cursor?

File: cmd/entire/cli/agent/cursor/lifecycle.go [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+	"fmt"
+	"io"
+	"os"
+	"time"
+
+	"github.com/entireio/cli/cmd/entire/cli/agent"
+)
+
+// HookNames returns the hook verbs Cursor supports.
+// Delegates to GetHookNames for backward compatibility.
```
Comment (Soph) [inline cmd/entire/cli/agent/cursor/lifecycle.go]: I guess this can be removed? there is no backward compatibility, right?

File: cmd/entire/cli/agent/cursor/hooks.go [hunk 8]
```diff
@@ -0,0 +1,379 @@
...
+
+	return count, nil
+}
+
+// UninstallHooks removes Entire hooks from Cursor HooksFileName.
+// Unknown top-level fields and hook types are preserved on round-trip.
+func (c *CursorAgent) UninstallHooks() error {
+	repoRoot, err := paths.RepoRoot()
+	if err != nil {
+		repoRoot = "."
```
Comment (Soph) [inline cmd/entire/cli/agent/cursor/hooks.go]: `InstallHooks()` falls back to `repoRoot, err = os.Getwd()`

File: cmd/entire/cli/agent/cursor/hooks.go [hunk 9]
```diff
@@ -0,0 +1,379 @@
...
+		return fmt.Errorf("failed to write "+HooksFileName+": %w", err)
+	}
+	return nil
+}
+
+// AreHooksInstalled checks if Entire hooks are installed.
+func (c *CursorAgent) AreHooksInstalled() bool {
+	repoRoot, err := paths.RepoRoot()
+	if err != nil {
+		repoRoot = "."
```
Comment (Soph) [inline cmd/entire/cli/agent/cursor/hooks.go]: `InstallHooks()` falls back to `repoRoot, err = os.Getwd()`

File: cmd/entire/cli/agent/cursor/cursor.go [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+	data, err := os.ReadFile(input.SessionRef)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read transcript: %w", err)
+	}
+
+	return &agent.AgentSession{
+		SessionID:  input.SessionID,
+		AgentName:  c.Name(),
+		SessionRef: input.SessionRef,
+		StartTime:  time.Now(),
```
Comment (Soph) [inline cmd/entire/cli/agent/cursor/cursor.go:108]: Does the session have no initial timestamp? Maybe that from the first event in it?

File: cmd/entire/cli/agent/cursor/hooks.go [hunk 11]
```diff
@@ -0,0 +1,379 @@
...
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/entireio/cli/cmd/entire/cli/agent"
+	"github.com/entireio/cli/cmd/entire/cli/jsonutil"
+	"github.com/entireio/cli/cmd/entire/cli/paths"
+)
+
+// Ensure CursorAgent implements HookSupport and HookHandler
```
Comment (Soph) [inline cmd/entire/cli/agent/cursor/hooks.go]:
```suggestion
// Ensure CursorAgent implements HookSupport
```

File: cmd/entire/cli/agent/cursor/cursor.go (hunk 10 above)
Comment (squishykid) [reply cmd/entire/cli/agent/cursor/cursor.go:108]: I think this is a hangover from the current agent interface usage. We never actually call `ReadSession` from from agent interface. We do also we never use AgentSession.StartTime. Claude also uses `time.Now()` for `startTime`.

Thinking out loud:
- we could use the create time for the file at `input.SessionRef`.
- we could remove the AgentSession.StartTime field

File: cmd/entire/cli/strategy/manual_commit_condensation.go
Comment (squishykid) [reply cmd/entire/cli/strategy/manual_commit_condensation.go:545]: good catch!

**Comment by squishykid** [comment]
@cursor review

File: cmd/entire/cli/agent/cursor/types.go [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+	Command string `json:"command"`
+	Matcher string `json:"matcher,omitempty"`
+}
+
+// sessionInfoRaw is the JSON structure from SessionStart/SessionEnd/Stop hooks.
+// Cursor may provide session_id or conversation_id (fallback).
+type sessionInfoRaw struct {
+	// common
+	ConversationID string   `json:"conversation_id"`
+	GenerationID   string   `json:"generation_id"`
```
[the remark this answers is not in the document]
Comment (squishykid) [reply cmd/entire/cli/agent/cursor/types.go:43]: good catch

File: cmd/entire/cli/agent/cursor/lifecycle.go [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+func (c *CursorAgent) parseSessionStart(stdin io.Reader) (*agent.Event, error) {
+	raw, err := agent.ReadAndParseHookInput[sessionInfoRaw](stdin)
+	if err != nil {
+		return nil, err
+	}
+	return &agent.Event{
+		Type:       agent.SessionStart,
+		SessionID:  raw.ConversationID,
+		SessionRef: raw.TranscriptPath,
+		Timestamp:  time.Now(),
```
[the remark this answers is not in the document]
Comment (squishykid) [reply cmd/entire/cli/agent/cursor/lifecycle.go:59]: session_id and conversation_id are the same thing in cursor. we use conversation_id because it is available on every hook.

File: cmd/entire/cli/agent/cursor/types.go [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+	Summary              string      `json:"summary"`
+	ParentConversationID string      `json:"parent_conversation_id"`
+	MessageCount         json.Number `json:"message_count"`
+	ToolCallCount        json.Number `json:"tool_call_count"`
+	ModifiedFiles        []string    `json:"modified_files"`
+	LoopCount            json.Number `json:"loop_count"`
+	Task                 string      `json:"task"`
+	Description          string      `json:"description"`
+	AgentTranscriptPath  string      `json:"agent_transcript_path"`
+}
```
[the remark this answers is not in the document]
Comment (squishykid) [reply cmd/entire/cli/agent/cursor/types.go:138]: marshaling / unmarshaling JSON from/to structs with an embedded struct is a pain.

File: cmd/entire/cli/agent/cursor/cursor.go (hunk 10 above)
Comment (Soph) [reply cmd/entire/cli/agent/cursor/cursor.go:108]: hmm, ok but also thinking more: this is triggered when the hook comes in at the start of a session. So I guess thinking about this today again: `time.Now()` is good enough? Like it would be maximum a few seconds delay but that also shouldn't really matter? 🤔

Comment (squishykid) [reply cmd/entire/cli/agent/cursor/cursor.go:108]: I think I’ll make a follow up ticket to look into this a bit deeper. Especially since we use `time.now()` in existing agent implementations.

**Review by Soph — APPROVED** [review]
